File: open/hackenbushredblue.py
import sys
from functools import lru_cache
from collections import deque

it = iter(sys.stdin.read().split())
n, m = int(next(it)), int(next(it))

edges = [None] * m
color = [''] * m
adj_list = [[] for _ in range(n + 1)]
for i in range(m):
    x = int(next(it))
    y = int(next(it))
    c = next(it)[0]
    if x == -1: x = 0
    if y == -1: y = 0

    edges[i] = (x, y)
    color[i] = c
    adj_list[x].append((y, i))
    adj_list[y].append((x, i))

@lru_cache(None)
def prune(m0):
    visited = [False] * (n + 1)
    visited[0] = True
    q = deque([0])
    while q:
        v = q.popleft()
        for u, i in adj_list[v]:
            if m0 >> i & 1 and not visited[u]:
                visited[u] = True
                q.append(u)

    m1 = 0
    for i in range(m):
        if m0 >> i & 1 and visited[edges[i][0]] and visited[edges[i][1]]: m1 |= 1 << i
    return m1

@lru_cache(None)
def dfs(m0):
    if not m0: return 0, 0

    visited = [False] * (n + 1)
    comps = []
    for s in range(1, n + 1):
        if visited[s] or not any(m0 >> i & 1 for _, i in adj_list[s]): continue
        m1 = 0
        visited[s] = True
        q = deque([s])
        while q:
            v = q.popleft()
            for u, i in adj_list[v]:
                if m0 >> i & 1:
                    m1 |= 1 << i
                    if u and not visited[u]:
                        visited[u] = True
                        q.append(u)

        comps.append(m1)
    if len(comps) > 1:
        p0, q0 = 0, 0
        for m1 in sorted(comps):
            p1, q1 = dfs(m1)
            if q0 < q1:
                p0 <<= q1 - q0
                q0 = q1
            else: p1 <<= q0 - q1
            p0 += p1
        return p0, q0

    x = y = False
    px = qx = py = qy = 0
    m1 = m0
    while m1:
        lsb = m1 & -m1
        m1 ^= lsb
        p, q = dfs(prune(m0 ^ lsb))

        def cmp(p1, q1, p2, q2):
            if q1 < q2: p1 <<= q2 - q1
            elif q1 > q2: p2 <<= q1 - q2
            return p1 - p2

        if color[lsb.bit_length() - 1] == 'B':
            if not x or cmp(p, q, px, qx) > 0:
                px, qx = p, q
                x = True
        else:
            if not y or cmp(p, q, py, qy) < 0:
                py, qy = p, q
                y = True

    def scale(p, q, Q):
        if q >= Q: return p >> q - Q
        return p << Q - q

    if x and y:
        l, r = -1, max(qx, qy) + 1
        while l + 1 < r:
            m = l + (r - l) // 2
            if scale(px, qx, m) + 1 < -scale(-py, qy, m): r = m
            else: l = m
        return scale(px, qx, r) + 1, r
    if x:
        if px < 0: return 0, 0
        return scale(px, qx, 0) + 1, 0
    if y:
        if py > 0: return 0, 0
        P = scale(py, qy, 0)
        return P - 1 if not qy and P == py else P, 0
    return 0, 0

P, Q = dfs(prune((1 << m) - 1))
while Q and not P & 1:
    P >>= 1
    Q -= 1

print(P if not Q else f"{P}/2^{Q}")

File: open/hackenbushredblue.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    string color(m, '#');
    vector<pair<int, int>> edges(m);
    vector<vector<pair<int, int>>> adj_list(n + 1);
    for (int i = 0; i < m; i++) {
        auto &[x, y] = edges[i];
        char c;
        cin >> x >> y >> c;
        if (!~x) x = 0;
        if (!~y) y = 0;

        color[i] = c;
        adj_list[x].emplace_back(y, i);
        adj_list[y].emplace_back(x, i);
    }

    vector<int> memo1(1 << (m + 1), -1);
    auto prune = [&](int mask) -> int {
        if (~memo1[mask]) return memo1[mask];

        vector<bool> visited(n + 1, false);
        visited[0] = true;
        queue<int> q;
        q.emplace(0);
        while (!q.empty()) {
            int v = q.front();
            q.pop();

            for (auto [u, i] : adj_list[v])
                if ((mask >> i) & 1)
                    if (!visited[u]) {
                        visited[u] = true;
                        q.emplace(u);
                    }
        }

        memo1[mask] = 0;
        for (int i = 0; i < m; i++)
            if (((mask >> i) & 1) && visited[edges[i].first] && visited[edges[i].second]) memo1[mask] |= 1 << i;
        return memo1[mask];
    };

    vector<pair<long long, int>> memo2(1 << (m + 1), {-1, -1});
    auto dfs = [&](auto &&self, int m0) -> pair<long long, int> {
        if (memo2[m0] != make_pair(-1LL, -1)) return memo2[m0];
        if (!m0) return memo2[m0] = {0, 0};

        vector<bool> visited(n + 1, false);
        vector<int> component_masks;
        for (int s = 1; s <= n; s++) {
            if (visited[s] || !any_of(adj_list[s].begin(), adj_list[s].end(), [&](auto e) { return (m0 >> e.second) & 1; })) continue;

            int m1 = 0;
            visited[s] = true;
            queue<int> q;
            q.emplace(s);
            while (!q.empty()) {
                int v = q.front();
                q.pop();

                for (auto [u, i] : adj_list[v])
                    if ((m0 >> i) & 1) {
                        m1 |= 1 << i;
                        if (u && !visited[u]) {
                            visited[u] = true;
                            q.emplace(u);
                        }
                    }
            }
            component_masks.emplace_back(m1);
        }

        auto scale = [&](long long p, int q, int Q) -> long long {
            if (q >= Q) return p >> (q - Q);
            return p << (Q - q);
        };

        if (component_masks.size() > 1) {
            auto p0 = 0LL;
            int q0 = 0;
            for (int m1 : component_masks) {
                auto [p1, q1] = self(self, m1);
                int Q = max(q0, q1);
                p0 = scale(p0, q0, Q) + scale(p1, q1, Q);
                q0 = Q;
            }
            return memo2[m0] = {p0, q0};
        }

        auto cmp = [&](long long p1, int q1, long long p2, int q2) -> long long {
            int Q = max(q1, q2);
            return scale(p1, q1, Q) - scale(p2, q2, Q);
        };

        bool x = false, y = false;
        auto px = 0LL, py = 0LL;
        int qx = 0, qy = 0;
        for (int m1 = m0; m1;) {
            int lsb = m1 & -m1;
            m1 ^= lsb;
            auto [p, q] = self(self, prune(m0 ^ lsb));
            if (color[countr_zero((unsigned) lsb)] == 'B') {
                if (!x || cmp(p, q, px, qx) > 0) {
                    px = p;
                    qx = q;
                    x = true;
                }
            } else {
                if (!y || cmp(p, q, py, qy) < 0) {
                    py = p;
                    qy = q;
                    y = true;
                }
            }
        }

        if (x && y) {
            int l = -1, r = max(qx, qy) + 1, mid;
            while (l + 1 < r) {
                mid = l + (r - l) / 2;

                if (scale(px, qx, mid) + 1 < -scale(-py, qy, mid)) r = mid;
                else l = mid;
            }
            return memo2[m0] = {scale(px, qx, r) + 1, r};
        }

        if (x) {
            if (px < 0) return memo2[m0] = {0, 0};
            return memo2[m0] = {scale(px, qx, 0) + 1, 0};
        }

        if (y) {
            if (py > 0) return memo2[m0] = {0, 0};
            auto P = scale(py, qy, 0);
            return memo2[m0] = {P - (!qy && P == py), 0};
        }
        return memo2[m0] = {0, 0};
    };
    auto [P, Q] = dfs(dfs, prune((1 << m) - 1));
    for (; Q && !(P & 1); P >>= 1, Q--);

    if (!Q) cout << P;
    else cout << P << "/2^" << Q;
}
